src/neml2/user_tensors/MultiColumnCSVScalar.cxx

```diff
@@ -0,0 +1,155 @@
+// All Rights Reserved
+// Software Name: NEML2 -- the New Engineering material Model Library, version 2
+// By: Argonne National Laboratory
+// OPEN SOURCE LICENSE (MIT)
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#ifdef NEML2_HAS_CSV
+
+#include "neml2/user_tensors/MultiColumnCSVScalar.h"
+#include "csvparser/csv.hpp"
+
+namespace neml2
+{
+register_NEML2_object(MultiColumnCSVScalar);
+
+OptionSet
+MultiColumnCSVScalar::expected_options()
+{
+  OptionSet options = UserTensorBase::expected_options();
+  options.doc() = "Construct a Scalar from a CSV file by column.";
+
+  options.set<std::string>("csv_file");
+  options.set("csv_file").doc() = "Path to the CSV file";
+
+  options.set<TensorShape>("batch_shape");
+  options.set("batch_shape").doc() = "Batch shape";
+
+  options.set<std::vector<std::string>>("column_names") = {};
+  options.set("column_names").doc() = "Names of CSV columns.";
+
+  options.set<std::vector<unsigned int>>("column_indices") = {};
+  options.set("column_indices").doc() = "Indices of CSV columns.";
+
+  options.set<bool>("header_row") = true;
+  options.set("header_row").doc() = "Whether the CSV file has a header row. The default is true.";
+
+  return options;
+}
+
+MultiColumnCSVScalar::MultiColumnCSVScalar(const OptionSet & options)
+  : Scalar(parse_csv(options)),
+    UserTensorBase(options)
+{
+}
+
+Scalar
+MultiColumnCSVScalar::parse_csv(const OptionSet & options) const
+{
+  std::string csv_file = options.get<std::string>("csv_file");
+  TensorShape batch_shape = options.get<TensorShape>("batch_shape");
+  std::vector<std::string> column_names = options.get<std::vector<std::string>>("column_names");
+  std::vector<unsigned int> column_indices =
+      options.get<std::vector<unsigned int>>("column_indices");
+  bool header_row = options.get<bool>("header_row");
+
+  // Allow the following delimiters
+  csv::CSVFormat format;
+  format.delimiter({',', ' ', ';', '\t'});
```

Comment:
I think it might be a better idea to allow one and only one delimiter at a time. What do you think? You can make the choice of delimiter an input option and default to comma.